docs/design_proposals/port-forward.md

```diff
@@ -0,0 +1,145 @@
+by @corneliusweig in https://github.com/GoogleContainerTools/skaffold/issues/1564#issuecomment-473528574.
+These changes are intended to be backwards-compatible with the previous version
+of Skaffold except where noted.
+
+
+### Command-Line Changes
+
+> **NOTE**: Ideally we would re-purpose Skaffold's existing `--port-forward` CLI argument in
+> a backward-compatible manner (still under investigation).  This document assumes it is
+> not possible.
+
+~Skaffold's `--port-forward` argument should be changed from a binary true/false option to~
+
+Skaffold will introduce a new `--port-forward-modes` argument that takes
+a set of comma-separated values with the following defined modes:
+
+   - `user`: user-defined port-forwards as defined in the `skaffold.yaml`
+   - `services`: service ports are forwarded
+   - `pods`: `containerPort`s defined on Pods and Kubernetes workload objects that have pod-specs
+     are forwarded (Deployment, ReplicaSet, StatefulSet, DaemonSet, Job, CronJob)
+   - `debug`: an internal strategy to forward debugging-related ports on Pods as set up
+     from `skaffold debug`
+   - `off`: no ports are ever forwarded
+
+#### Changes to port-forwarding defaults
+
+Skaffold will default to enabling port-forwarding for establishing user-defined port-forwards
+as defined in the `skaffold.yaml`.  Skaffold's `dev` and `debug` will be changed to the following defaults.
+
+Command-line                        | v1.15.0 default modes          | New default modes
+----------------------------------- | ---------------- | -------------------
+`skaffold dev`                        | off            | user
+`skaffold dev --port-forward`         | user, services | user, services (no change)
+`skaffold dev --port-forward=false`   | off            | off (no change)
+`skaffold debug`                      | off            | user, debug
+`skaffold debug --port-forward`       | user, services, pods | user, services, debug
+`skaffold debug --port-forward=false` | off            | off (no change)
+
+**Compatibility Change:** The behaviour of `skaffold debug --port-forward` no longer forwards all
+`containerPort`s defined on pods.  Container ports were forwarded as there was no other option to forward
+just debug ports.  With this proposed change, the `debug` mode will only select debug ports.  Forwarding all
+`containerPort`s was the cause of some confusion in earlier iterations of port-forwarding.
+
+
+### Skaffold configuration changes
+
+#### skaffold.yaml
+
+We will define a new configuration block `defaults` with a single array field `portForwardModes`:
+```yaml
+apiVersion: skaffold/v2beta8
+kind: Config
+build:
+  artifacts:
+  - image: foo
+defaults:
+  portForwardModes: ["user", "services"]
+```
+
+#### Global config ($HOME/.skaffold/config)
+
+The global config will support a `portForwardModes` with a single array field.  This can be set as a
```

Comment:
does persisting this in the global config mean that `dev` and `debug` will default to it in the absence of an explicit flag for every project? I'm not sure if that's a good thing. Doesn't seem like `portForwardModes` is a global (shared across projects sort of) variable like kubecontext or default repo.

Reply:
It could also be set on a per-context basis within the `~/.skaffold/config`.  The global default is good for people who are adamant that they never ever want port-forwarding.

Reply:
IMHO it's better to preserve the defaults for every new project. There's already a way to set it globally via the generated env variable `SKAFFOLD_PORT_FORWARD_MODES`. And all the other properties stored in global config inherently make sense persisted across projects. Is there any issue created by the _adamant user_?

Reply:
#1564 has a few comments.  At that time we forwarded all containers across all pods, so it could be confusing.  With saner defaults, this shouldn't be an issue, but it is worth implementing here.